lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1991,6 +2025,18 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       DC = DC->getParent();
     }
   }
+  
+  // Verify E is actually a descendant of ACE before we use it.
+  // Otherwise we'd be diagnosing E within the context of a closure
+  // that it isn't actually a part of.
+  if (ACE) {
+    auto isDescendantWalker = ExprIsDescendantWalker(const_cast<Expr *>(E));
+    ACE->walk(isDescendantWalker);
+    if (!isDescendantWalker.exprIsDescendant) {
+      ACE = nullptr;
```

Comment:
The implementation of `diagnoseImplicitSelfUseInClosure` effectively requires `E` to be a part of the body of `ACE`. 

For some reason, in this specific case using result builders, this code was being invoked with an expression that wasn't actually a descendant of the closure.

A fix that seems to work well here is to just check whether or not `E` is a descendant of `ACE` by walking the AST of the closure and checking if we encounter `E`.